### PR TITLE
Add get native handle

### PR DESCRIPTION
**Pull request by BastiaanOlij**
This is an alternative to #41137 and introduces a method called `get_native_handle` to our OS class. It is only exposed to GDScript so it is exposed to GDNative where it is really needed. This is mainly targeted at OpenXR support but will also benefit other solutions that require access to some of the platform internals to work. 

Currently it only implements WINDOW_VIEW (HDC) and OPENGL_INSTANCE (HGLRC) on the Windows platform which is why it's in draft.

We'll need similar implementations for Android (for which currently a custom gdnative API exists that could potentially be retired) and X11, again to support OpenXR and other AR/VR platforms. 

Note, currently focusing on 3.2 but this will be vital to XR on Godot 4 in exposing parts of Vulkan. Note that the index added is meant for Godot 4's multiple window support. I will be adding documentation to this.

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
cc @bruvzg , @m4gr3d , @ChristophHaag

**Review by bruvzg — APPROVED** [review]

**Comment by BastiaanOlij** [comment]
I've added the other options to the Windows version of get_native_handle and implemented an (untested) linux version. 
@ChristophHaag could you check if what I did makes sense?

File: platform/x11/os_x11.cpp [hunk 1]
```diff
@@ -1756,6 +1756,18 @@ void OS_X11::request_attention() {
 	XFlush(x11_display);
 }
 
+void *OS_X11::get_native_handle(int p_handle_type, int p_index) {
+	/* Godot 3.2 does not have multiple window support so we ignore p_index */
+	switch (p_handle_type) {
+		case APPLICATION_HANDLE: return NULL; // Do we have a value to return here?
+		case DISPLAY_HANDLE: return &x11_display;
+		case WINDOW_HANDLE: return &x11_window;
```
Comment (bruvzg) [inline platform/x11/os_x11.cpp:1764]: Any reason to use addresses? `x11_display` is already a pointer type (X11::Display *) and `x11_window` is `XID` == `unsigned long`, both should fit by value.

File: core/bind/core_bind.h [hunk 2]
```diff
@@ -143,6 +143,14 @@ class _OS : public Object {
...
 	};
 
+	enum HandleType {
+		APPLICATION_HANDLE, // HINSTANCE, NSApplication*, UIApplication*, JNIEnv* ...
+		DISPLAY_HANDLE, // X11::Display* ...
+		WINDOW_HANDLE, // HWND, X11::Window*, NSWindow*, UIWindow*, Android activity ...
+		WINDOW_VIEW, // HDC, NSView*, UIView*, Android surface ...
+		OPENGL_CONTEXT // HGLRC, X11::GLXContext, NSOpenGLContext*, EGLContext* ...
+	};
+
```
Comment (m4gr3d) [inline core/bind/core_bind.h:153]: Is it necessary to define this here since `core_bind.h` includes `os.h` which also contains a similar `HandleType` declaration?

File: platform/x11/os_x11.cpp (hunk 1 above)
Comment (BastiaanOlij) [reply platform/x11/os_x11.cpp:1764]: I'll change it, I wasn't sure because I don't do much with Linux so I did that one blind. We are returning a signed int because I think that is what GDNative / Variant type support but hopefully that will not pose a problem.

File: core/bind/core_bind.h (hunk 2 above)
Comment (BastiaanOlij) [reply core/bind/core_bind.h:153]: I think so, I'm not 100% sure but the same is done with a number of other enums.. The problem is that core_bind is a wrapper that exposes OS to GDScript/GDNative and we can only expose constants if they are part of the class being exposed. 

I could be wrong though, I should experiment with it a little more.

File: core/bind/core_bind.h [hunk 3]
```diff
@@ -143,6 +143,14 @@ class _OS : public Object {
 		MONTH_DECEMBER
 	};
 
+	enum HandleType {
+		APPLICATION_HANDLE, // HINSTANCE, NSApplication*, UIApplication*, JNIEnv* ...
+		DISPLAY_HANDLE, // X11::Display* ...
+		WINDOW_HANDLE, // HWND, X11::Window*, NSWindow*, UIWindow*, Android activity ...
+		WINDOW_VIEW, // HDC, NSView*, UIView*, Android surface ...
+		OPENGL_CONTEXT // HGLRC, X11::GLXContext, NSOpenGLContext*, EGLContext* ...
```
Comment (akien-mga) [inline core/bind/core_bind.h:151]:
```suggestion
		OPENGL_CONTEXT, // HGLRC, X11::GLXContext, NSOpenGLContext*, EGLContext* ...
```
Prevent having a diff on that line next time we add enum values.

File: core/os/os.h [hunk 4]
```diff
@@ -238,6 +238,22 @@ class OS {
...
+	// Note that whether a valid handle is returned depends on whether it applies to the given
+	// platform and often to the chosen render driver.
+	// NULL will be returned if a handle is not available.
+
+	enum HandleType {
+		APPLICATION_HANDLE, // HINSTANCE, NSApplication*, UIApplication*, JNIEnv* ...
+		DISPLAY_HANDLE, // X11::Display* ...
+		WINDOW_HANDLE, // HWND, X11::Window*, NSWindow*, UIWindow*, Android activity ...
+		WINDOW_VIEW, // HDC, NSView*, UIView*, Android surface ...
+		OPENGL_CONTEXT // HGLRC, X11::GLXContext, NSOpenGLContext*, EGLContext* ...
```
Comment (akien-mga) [inline core/os/os.h:252]:
```suggestion
		OPENGL_CONTEXT, // HGLRC, X11::GLXContext, NSOpenGLContext*, EGLContext* ...
```

File: core/bind/core_bind.cpp [hunk 5]
```diff
@@ -1292,6 +1297,8 @@ void _OS::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("center_window"), &_OS::center_window);
 	ClassDB::bind_method(D_METHOD("move_window_to_foreground"), &_OS::move_window_to_foreground);
 
+	ClassDB::bind_method(D_METHOD("get_native_handle"), &_OS::get_native_handle);
```
Comment (akien-mga) [inline core/bind/core_bind.cpp:1300]: You should give the argument names too (that's why currently you have `arg0` and `arg1` in the doc).

Comment (akien-mga) [reply core/bind/core_bind.cpp:1300]:
```suggestion
	ClassDB::bind_method(D_METHOD("get_native_handle", "handle_type", "index"), &_OS::get_native_handle);
```

File: doc/classes/OS.xml [hunk 6]
```diff
@@ -1058,7 +1070,7 @@
 		<member name="screen_orientation" type="int" setter="set_screen_orientation" getter="get_screen_orientation" enum="_OS.ScreenOrientation" default="0">
 			The current screen orientation.
 		</member>
-		<member name="tablet_driver" type="String" setter="set_current_tablet_driver" getter="get_current_tablet_driver" default="&quot;&quot;">
+		<member name="tablet_driver" type="String" setter="set_current_tablet_driver" getter="get_current_tablet_driver" default="&quot;wintab&quot;">
```
Comment (akien-mga) [inline doc/classes/OS.xml:1073]: That's a quirk of `doctool` which sometimes have platform-specific behavior, but you should revert this change manually.

File: doc/classes/OS.xml [hunk 7]
```diff
@@ -1162,6 +1174,21 @@
 		<constant name="MONTH_DECEMBER" value="12" enum="Month">
 			December.
 		</constant>
+		<constant name="APPLICATION_HANDLE" value="0" enum="HandleType">
+			Application handle: HINSTANCE, NSApplication*, UIApplication*, JNIEnv* ...
```
Comment (akien-mga) [inline doc/classes/OS.xml:1178]: All these should use `[code]` markup. Might need some more explanations about what each handle is and can be used for. (And then yeah a list of possible handle types depending on the platform.)

File: platform/x11/os_x11.cpp [hunk 8]
```diff
@@ -1756,6 +1756,18 @@ void OS_X11::request_attention() {
 	XFlush(x11_display);
 }
 
+void *OS_X11::get_native_handle(int p_handle_type, int p_index) {
+	/* Godot 3.2 does not have multiple window support so we ignore p_index */
```
Comment (akien-mga) [inline platform/x11/os_x11.cpp:1760]: Then maybe `p_index` doesn't need to be added in the 3.2 version of this PR?

File: doc/classes/OS.xml (hunk 7 above)
Comment (akien-mga) [reply doc/classes/OS.xml:1178]: The docs should likely also list exactly what is returned on each platform, instead of potential handles that might be returned in the future, so that users know what they'll get without having to read the Godot source code.

File: platform/x11/os_x11.cpp (hunk 8 above)
Comment (BastiaanOlij) [reply platform/x11/os_x11.cpp:1760]: Yeah I've been wondering whether we should leave that out and then just add it back in when I do the PR for Godot 4

File: doc/classes/OS.xml (hunk 6 above)
Comment (BastiaanOlij) [reply doc/classes/OS.xml:1073]: Will do tomorrow :)

File: core/bind/core_bind.cpp (hunk 5 above)
Comment (BastiaanOlij) [reply core/bind/core_bind.cpp:1300]: Ah! I was wondering about that, too used to not doing that in GDNative plugins

File: core/bind/core_bind.h (hunk 2 above)
Comment (BastiaanOlij) [reply core/bind/core_bind.h:153]: Actually @akien-mga , can you confirm this? Do enum defs have to be in the same class for the bind macros to work?

Comment (akien-mga) [reply core/bind/core_bind.h:153]: I think so, it's probably not possible to bind enums which are not defined in the current class as part of that class' API.
Might have improved with @reduz's recent refactoring of method binding, not sure.

**Comment by BastiaanOlij** [comment]
Applied all of @akien-mga requests (I think), I'm only not sure how I can get rid of the blank lines in the documentation:
image

**Comment by akien-mga** [comment]
>  I'm only not sure how I can get rid of the blank lines in the documentation:
image

You can't for now, there's no `[ul]` implemented in the docs, and each line is treated as a paragraph.

File: doc/classes/OS.xml [hunk 9]
```diff
@@ -1162,6 +1172,34 @@
 		<constant name="MONTH_DECEMBER" value="12" enum="Month">
 			December.
 		</constant>
+		<constant name="APPLICATION_HANDLE" value="0" enum="HandleType">
+			Application handle:
```
Comment (akien-mga) [inline doc/classes/OS.xml:1176]: You have a trailing space on each first line of the constant descriptions.

File: core/bind/core_bind.h (hunk 2 above)
Comment (BastiaanOlij) [reply core/bind/core_bind.h]: Well that would be Godot 4 anyway, so we could revisit it when we port this to 4

File: doc/classes/OS.xml (hunk 9 above)
Comment (BastiaanOlij) [reply doc/classes/OS.xml]: yeah I noticed, already fixed :)

**Comment by BastiaanOlij** [comment]
> 
> You can't for now, there's no `[ul]` implemented in the docs, and each line is treated as a paragraph.

So we'll just leave it like this for now?

**Comment by akien-mga** [comment]
> So we'll just leave it like this for now?

Yeah that's OK. When someone ends up being too bothered by this bad looking lists, they'll figure out a way to improve the visual output :P

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
